\documentclass[english]{article}

%%%%%%%%%%%%%%%%
%---PACKAGES---%
%%%%%%%%%%%%%%%%
\usepackage[english]{babel}
\usepackage[T1]{fontenc}
\usepackage[latin9]{inputenc}
\usepackage{natbib}
\usepackage{url}
\usepackage{float}
\usepackage{graphicx}
\usepackage{dcolumn}
\usepackage{multirow}
\usepackage{booktabs}
\usepackage{color}
\usepackage{multirow}
\usepackage{amsmath}
\usepackage{amssymb}
\usepackage{textpos}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%---USER SPECIFIED LaTeX COMMANDS---%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%


% dcolumn definition of decimally aligned table
\newcolumntype{.}{D{.}{.}{-1}}

% Text layout (margins)
\topmargin 0.0cm
\oddsidemargin 0.5cm
\evensidemargin 0.5cm
\textwidth 16cm 
\textheight 21cm

% double line spacing
\linespread{1.6}

% paragraph indent
\setlength{\parindent}{0.5in}

% font for math
\usepackage{mathpazo}
\DeclareSymbolFont{letters}{OML}{ztmcm}{m}{it}
\DeclareSymbolFontAlphabet{\mathnormal}{letters}

\definecolor{fgcolor}{rgb}{0.2, 0.2, 0.2}


%%%%%%%%%%%%%%%%
%---DOCUMENT---%
%%%%%%%%%%%%%%%%
\begin{document}

% consistent edits between manuscript and rebuttal letter
\input{thesis.xtr}

\setcounter{page}{1}

\newcommand{\myTitle}{Continuous-time Markov chain models for pathogen phylodynamics: }
\newcommand{\mySubtitle}{model extension, simulation and visualisation} % produkcja dystrybucja masturbacja :)

\bigskip
\medskip
\noindent{\Large 
Response to the reviewer comments to: \myTitle \mySubtitle.
}
\bigskip

%TODO: I'm including only reviewers which had some major/minor comments

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%--- COMMENTS BY STIJN VANSTEELANDT---%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\section*{Comments by Stijn Vansteelandt}

\begin{itemize}

% 1
\item {
{\it
\textbf{
The thesis starts with a very nice introduction, which I think is very insightful for the reader who is not familiar with phylogenetic and/or Bayesian inference. 
In spite of this, I believe that the readability of the later chapters would be further improved if the introduction:
}% END: textbf
}% END: it
}% END: item

\begin{itemize}

% 1a
\item {
{\it
\textbf{
Briefly described the evolutionary models, such as GY94, that are later used.
}% END: textbf
}% END: it
}% END: item
\\
I have now added a \emph{Markov models of codon evolution} subsection in the introductory chapter, which discusses the most popular codon model implementations. The first paragraph of this new section is reproduced below (references are not reproduced correctly using the Latex `copy-over' tools we use in this response document):

\begin{quote}
\myeditsvonea
\end{quote}

% 1b
\item {
{\it
\textbf{
Briefly described how to accommodate time-stamped sequences, which is now considered in the analyses in Chapter 2, but not really discussed.
}% END: textbf
}% END: it
}% END: item
\\
%
%PL: TODO: rewrite the final paragraph to reflect the dated-tip likelihood model: Essential, we now consider heights (instead of branch lengths) and introduce a new parameter, the rate of evolution r, which allows rescaling genetic distance into time. The task is now to estimate the heights and r from the sequence data, keeping the the heights proportional to their sampling time, with the constant of proportionality being our estimable r.
%
I added a subsection which describes the methods used in BEAST to accommodate time-stamped sequences as a source for calibrating trees, when such information is available. The first paragraph of this new section is reproduced below:

\begin{quote}
\myeditsvoneb
\end{quote}

% 1c
\item {
{\it
\textbf{
Give a bit more detail on phylogeography, e.g. how do the models look like and what is the connection between phylogeography and epoch models (some connection is described in Chapter 2, but remains a bit vague).
}% END: textbf
}% END: it
}% END: item
\\
I added a figure and additional explanation, which should improve the flow of subsequent chapters. 
These changes start with:

\begin{quote}
\myeditsvonec
\end{quote}

\end{itemize}

% 2
\item {
{\it
\textbf{
It is sometimes unclear in Chapter 2 whether the analysis is performed under a fixed evolutionary tree or not. 
This is especially so in the evaluation of Bayes Factors in Table 2.3. 
It would also have been nice to understand better whether the use of the epoch model (versus not) has an essential impact on tree reconstruction. 
Somewhat related, it would have been nice to learn from the simulation study how mislead one could be if the data were generated under an epoch model, but analyzed under a simpler model. 
I am not asking for additional analyses here, but merely for additional clarification or insight.
}% END: textbf
}% END: it
}% END: item
\\
% FB: I don't get the first part of this comment. In the Intro we clearly write:
% "Our Bayesian approach also does not condition on a fixed tree topology but averages over all plausible evolutionary histories." (p 46) \\
% PL: OK, but to make it even more explicit I have added this to the results section (in simulation results): 'In all our analyses of the replicate data, as well as in the analyses of the real-world data sets in the next sections, we consider the tree as random and estimate it from the nucleotide data.' (this still needs to be tagged).
To re-iterate the fact that our model does not condition on a known tree topology we now write:

\begin{quote}
\myeditsvtwo
\end{quote}

We now also performed a number of additional simulations in order to explore the impact of correcting for time-heterogeneity on tree reconstruction.
The results of these simulations can serve as a basis for discussion during the thesis defence.
%The simulations which address the second part of this comment are now being performed, I will discuss them during the public defence.


% 3
\item {
{\it
\textbf{
P6, lines 5-7: this is not correctly phrased as the parameter values do not have an associated probability in the frequentist likelihood framework.
}% END: textbf
}% END: it
}% END: item
\\
This paragraph is now rephrased:

\begin{quote}
\myeditsvthree
\end{quote}


% 4
\item {
{\it
\textbf{
Expression (1.12) is confusing in that indices i and j appear in the lefthand side, but not in the righthand side. 
Likewise in expression (1.13).
}% END: textbf
}% END: it
}% END: item
\\
Corrected.



% 5
\item {
{\it
\textbf{
P20, 2nd display: an equality sign is missing on the 2nd line.
}% END: textbf
}% END: it
}% END: item
\\
Corrected.

% 6
\item {
{\it
\textbf{
I would find it useful to see a brief discussion on the plausibility of (1.22).
}% END: textbf
}% END: it
}% END: item
\\
% FB: Does he mean plausibility from a biological perspective or statistical? It's a necessary condition for the existence of stationary dist.
% Pl: I see it as questioning the plausibility in a biological sense. We now already argue that it is an assumption taken for computational reasons, but at the end of that paragraph we can explicitly state 'Realistic substitution processes are however unlikely to be time-reversible and procedures have also been developed to compute phylogenetic likelihoods under non-reversible models (cite Boussau and Gouy, 2006, see below)'. ...perhaps add in something more from that paper.... and end with 'We note that we use such non-reversible models in our discrete phytogeographic applications (cite Edwards et al., Current Biology, 2011).
% http://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.324.620&rep=rep1&type=pdf
Please see my changes which explain how abandoning the assumptions of stationarity and homogeneity of stochastic processes of evolution affect each other and which start with:

\begin{quote}
\myeditsvsixone
\end{quote}

Relaxing both of these assumptions leads to processes which are also non-reversible, and addition to explaining the implications of the time-reversibility assumption from biological perspective, we also briefly introduce some of the recent attempts to relax this assumption, both in the Bayesian phylogeographic framework and more broadly in the context of phylogenetic applications:

\begin{quote}
\myeditsvsixtwo
\end{quote}



% 7
\item {
{\it
\textbf{
If I am right that matrix P(t) on the bottom of page 22 is four-dimensional, then why not write it in full, which will be less confusing?
}% END: textbf
}% END: it
}% END: item
\\
Corrected.




% 8
\item {
{\it
\textbf{
P25, the statement at the end of the penultimate paragraph (`most plausible hypothesis for the data') is vague.
}% END: textbf
}% END: it
}% END: item
\\
This statement now reads:

\begin{quote}
\myeditsveight
\end{quote}



% 9
\item {
{\it
\textbf{
P28, I understand what T(n) is, but would nevertheless find it helpful if it were stated more explicitly.
}% END: textbf
}% END: it
}% END: item
\\
% FB: I don't understand the question in the definition we clearly define T as being simply a function on a positive subset of real numbers?
% Does he mean we should wrote "time"? but this is not always true?
% PL: OK, but can we then state something like this: T(n) can be interpreted as time in most practical situations, but this does necessarily need to be the case, e.g.....?
We now clarified that T(n) represents the ``worst-case' time' complexity of an algorithm.


% 10
\item {
{\it
\textbf{
P31, If the phrase `proper prior' is used, then perhaps it should be defined.
}% END: textbf
}% END: it
}% END: item
\\
This paragraph now reads:

\begin{quote}
\myeditsvnine
\end{quote}


% 11
\item {
{\it
\textbf{
P32, Since the author assumes that the reader may have no prior knowledge of Bayesian inference, I believe the meaning of Figure 1.11 should be explained in more detail.
}% END: textbf
}% END: it
}% END: item
\\
Corrected. There is now a more elaborate description accompanying this figure as well as the derivation of the results, starting with:

\begin{quote}
\myeditsveleven
\end{quote}

% 12
\item {
{\it
\textbf{
P33,. I think the notation $\theta[t]$ has not been defined.
}% END: textbf
}% END: it
}% END: item
\\
Corrected.




% 13
\item {
{\it
\textbf{
There is a typo in the lefthand side of (1.37).
}% END: textbf
}% END: it
}% END: item
\\
Corrected.


% 14
\item {
{\it
\textbf{
I would find it useful to get a bit more detail how a model for the likelihood of Y and for the prior of $\Phi$ might look like.
}% END: textbf
}% END: it
}% END: item
\\
% FB: I'm not sure if I understand this comment, we do write that the likelihood is standard tree pruning and there's a whole subsection in the Introduction that delves on that...  We could elaborate on the prior choice here  ...
% PL: OK, I would simply point out that we use exactly the same type of CTMC modelling in the response letter. And simply to do something in response to the comment, modify this sentence: "Unlike before, the rate matrix does not necessarily have to be symmetrical, i.e. there exist some i, j such that ..." into "So, we employ the same type of CTMC modelling for discrete phylogeography as for nucleotide substitution processes, which is described in detail in ...(refer to intro section on CTMC), but now for an arbitrary number of location states. And unlike for nucleotide substitution, the rate matrix does not necessarily have to be symmetrical, i.e. there exist some i, j such that ..." 
In response to the first comment raised by the Reviewer, we now provide more detail in Subsection~\ref{sub:phylogeo}, including a figure which complements the explanation of the joint inference of processes generating the sequence data and discrete location data.




% 15
\item {
{\it
\textbf{
P43. What is compositional heterogeneity?
}% END: textbf
}% END: it
}% END: item
\\
We now explicitly state that this refers to different stationary frequencies for the CTMC.

\begin{quote}
\myeditsvfifteen
\end{quote}


% 16
\item {
{\it
\textbf{
P44. I would find it useful to receive a bit more guidance concerning the connection to phylogeographic inference. 
Also in Section 2.4.3, the description of the proposal by Bahl et al. (2011) is too brief for me to appreciate why the epoch model is more appropriate.
}% END: textbf
}% END: it
}% END: item
\\
% FB: This becomes p46 now, from "The former is critical to accommodate discrete phylogeographic inference"
In addition to the detail added to the phylogeography section in the Introduction, I also added a sentence to clarify the approach taken by Bahl et al.:

\begin{quote}
\myeditsvsixteen
\end{quote}

This further clarifies the distinction between the \emph{ad-hoc} method by Bahl et al. and the ``epoch model'' applied to the global influenza spread problem.

% 17
\item {
{\it
\textbf{
Display (2.5): I believe this display implicitly assumes that the time T1 is a priori known. 
It would be good to state this explicitly.
}% END: textbf
}% END: it
}% END: item
\\
This is now stated explicitly.
%PL: where?may be quote here?


% 18
\item {
{\it
\textbf{
P57. The sentence `To determine the posterior odds, ... of that model' is vague.
}% END: textbf
}% END: it
}% END: item
\\
The sentence is rephrased to:

\begin{quote}
\myeditsveighteen
\end{quote}


% 19
\item {
{\it
\textbf{
P85. Riemannian -> Riemann
}% END: textbf
}% END: it
}% END: item
\\
Corrected.


% 20
\item {
{\it
\textbf{
P86. The last sentence of Section 5.1.3 is not sufficiently clear. 
Please make clear why the use of mixed effects models could allow this.
}% END: textbf
}% END: it
}% END: item
\\
I changed the relevant subsection, %PL: why not quote it (or at least part of it)?
explaining that by this sentence we mean the impact of various covariates,  and that by using the MG94 parametrisation we do not need to assume that all substitution rate variation is due variation in non-synonymous rates.
There is also a more elaborate description of the model we plan to use in the upcoming manuscript on the application of the hierarchical epoch model, starting with:

\begin{quote}
\myeditsvtwenty
\end{quote}


% 21
\item {
{\it
\textbf{
P88. It would be useful to receive a bit more detail what is meant by `infinite hidden Markov Model'.
}% END: textbf
}% END: it
}% END: item
\\
I now introduce the definition of HMM's and its extension to infinite states under the a form of Remark on page 93,
% ~\pageref{ref:infinite_hidden_markov}, 
please see the relevant subsection.


\end{itemize}


%%%%%%%%%%%%%%%%%%%%
%--- COMMENTS---%
%%%%%%%%%%%%%%%%%%%%
\section*{Comments by David Posada}

\begin{itemize}

% 1
\item {
{\it
\textbf{
I missed a brief primer of Bayesian model selection in the introductory chapter.
Techniques such Bayes factors, path sampling or stepping-stone sampling are used throughout but never explained. 
This should be very easy to fix.
}% END: textbf
}% END: it
}% END: item
\\
There is now a subsection called \emph{Bayesian model selection} at the end of Chapter 1, which gives a primer on Bayesian model selection that should be sufficient to follow the rest of the manuscript.



% 2
\item {
{\it
\textbf{
In the simulations in Chapter 2 you assume you know the true epoch times, but this information is rarely available in real life. 
Simulation 1 addresses the impact of model misspecification partially, as the true process does not have epochs. 
It could be a good idea to further check the impact of model misspecification, in particular simulating a number of epochs that do not correspond with the assumed ones (you later mention this idea in the final chapter but do not end up testing its potential biasing effect). 
Also, simulating under a non-epoch model - for example where different process can apply to contemporaneous branches, would be fundamental to understand the robustness of the model. 
This does not need to be done before the defense, but it could be a nice extension for the future.
}% END: textbf
}% END: it
}% END: item
\\
Indeed, the cases we examine in this thesis involve testing prior hypotheses with known epoch times.
We are performing a set of simulations to dress some of the points raised by the reviewer, and the results can serve as a basis for discussion at the public defence.


% 3
\item {
{\it
\textbf{
In general, there is some repetitive text, sometimes with identical phrases/paragraphs, between the introduction and some of the other chapters. 
On the other hand, some relevant aspects of the research are discussed in the general discussion in Chapter 5, instead of in the corresponding `research' chapter. 
This is true in particular for Chapter 3, were a very concise description of SPREAD is provided. 
I understand these redundancies and fragmentations are a byproduct of reproducing papers as chapters, but however these issues have been corrected in some occasions. 
For example, chapter 2 is not identical to the SysBio paper. 
I suggest you try a bit to further consolidate the different chapters.
}% END: textbf
}% END: it
}% END: item
\\
We carefully screened for repeated paragraphs between the introduction and the other chapters, and could only identify one paragraph that is repeated at the beginning of Subsection~\ref{sub:subst_models}.
This paragraph is now rephrased to remove the repetition.

We acknowledge that isolated sentences might still carry some degree of resemblance, yet we prefer not to remove this so that all chapters remain as self-contained units, for readers that would only be interested in parts of the thesis.
For Chapter 3, we now clearly reference the SPREAD tutorial that we provide as an appendix, which contains the nuts and the bolts of the software tool that are important for the practitioner.
This complements the chapter but would be difficult to merge this with the original paper describing SPREAD.



% 4
\item {
{\it
\textbf{
The simulation software piBUSS described in Chapter 4 can be very useful, in particular in regards to its connectivity with BEAST. 
The possibility of simulating discrete phylogeographic traits is very convenient. 
Here it would have been interesting to see, for the sake of comparison and interpretation, the correspondence between simulated and estimated tMRCAs when the true model has one epoch with omega=0.5 (or even better, omega equal to the weighted mean of the n-epoch model). 
This does not need to be done before the defense, but it could be a nice extension for the future.
}% END: textbf
}% END: it
}% END: item
\\
This is an interesting suggestion indeed. We are performing a set of simulations to asses this, and I would be happy to discuss this with the reviewer atthe public defence.


% 5
\item {
{\it
\textbf{
Page 14: I would say that the term `Big Data' does not apply to sequence alignments. 
It might do for NGS files etc., but not for the type of data that BEAST uses as input. 
Not yet. 
Moreover, you focus on HPC architectures like GPU and multicores, which are in some senses opposite to Big Data architectures (i.e., Hadoop Map Reduce).
}% END: textbf
}% END: it
}% END: item
\\
I have rephrased the entire paragraph to avoid using the term:

\begin{quote}
\myeditDPfive
\end{quote}


% 6
\item {
{\it
\textbf{
Page 25, 31: likelihood should be $P(X| \theta)$, not $P(\theta | X)$.
}% END: textbf
}% END: it
}% END: item
\\
Corrected.


% 7
\item {
{\it
\textbf{
Page 34: on the left-term of equation 1.37, the denominator should be
$P(\theta[t]|X) q(\theta^{*} | \theta[t])$
not 
$P(\theta[t]|X) q(\theta[t] | \theta^{*})$
}% END: textbf
}% END: it
}% END: item
\\
Corrected.


% 8
\item {
{\it
\textbf{
Page 52: according to the legend in Figure 2.3, a two-epoch plot is missed.
}% END: textbf
}% END: it
}% END: item
\\
The complete specification of the two simulation scenarios is presented in one plot by using two different epoch delineations (at the top and bottom of the tree) 
and alternating shading, with one shaded area having a transparency gradient.
This was perhaps not clear enough in the Figure caption, I have thus rephrased more explicitly:

\begin{quote}
{\bf  Epoch simulation scenarios on an influenza A maximum clade credibility tree topology.} 
% \myedit{DPeight}{
In the two-epoch example illustrated at the top with alternating grey and light grey background
% at the top, 
the transition time is set at $t_{1}=7$, creating two epochs with substitution processes governed by infinitesimal rate matrices $\mathbf{Q_{1}}$ and $\mathbf{Q_{2}}$ respectively.
The light area (one of which encompasses  two epochs in the 3-epoch scenario as indicated by the transparency gradient for the leftmost epoch) is separated from the dark grey area by a dotted line.
In the three-epoch example illustrated 
at the bottom, % at the bottom, 
the transition times are put at $t_{1}=7$ and $t_{2}=15$, creating three epochs with substitution processes governed by infinitesimal rate matrices $\mathbf{Q_{1}}$, $\mathbf{Q_{2}}$ and then again $\mathbf{Q_{1}}$, as indicated by the alternating dark and light grey areas.
% }%END: myedit
\end{quote}



% 9
\item {
{\it
\textbf{
The bibliography would have been be easier to consult if last names came first than the initials, as usual.
}% END: textbf
}% END: it
}% END: item
\\
Corrected.


% 10
\item {
{\it
\textbf{
Page 137: Figure B.12 should be in color. 
Otherwise it is very difficult to grasp.
}% END: textbf
}% END: it
}% END: item
\\
Corrected, all the figures presenting SPREAd output are now in color.

% 11
\item {
{\it
\textbf{
Page 140: Figure B.13 should be in color. 
Otherwise it is very difficult to grasp.
}% END: textbf
}% END: it
}% END: item
\\
Corrected, all the figures presenting SPREAD output are now in color.


% 12
\item {
{\it
\textbf{
Page 143: Figure B.15 should be in color. 
Otherwise it is very difficult to grasp.
}% END: textbf
}% END: it
}% END: item
\\
Corrected, all the figures presenting SPREAD output are now in color.




% 13
\item {
{\it
\textbf{
Unless I missed it, the appendices are not cited in the text.
}% END: textbf
}% END: it
}% END: item
\\
Thank you for catching this oversight. We now refer to the Appendices in the Chapters to which they apply.


\end{itemize}





\bibliographystyle{myabbrvnat.bst}
\bibliography{thesisrefs}

\end{document}
