\documentclass[english]{article}

%%%%%%%%%%%%%%%%
%---PACKAGES---%
%%%%%%%%%%%%%%%%
\usepackage[english]{babel}
\usepackage[T1]{fontenc}
\usepackage[latin9]{inputenc}
\usepackage{natbib}
\usepackage{url}
\usepackage{float}
\usepackage{graphicx}
\usepackage{dcolumn}
\usepackage{multirow}
\usepackage{booktabs}
\usepackage{color}
\usepackage{multirow}
\usepackage{amsmath}
\usepackage{amssymb}
\usepackage{textpos}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%---USER SPECIFIED LaTeX COMMANDS---%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%


% dcolumn definition of decimally aligned table
\newcolumntype{.}{D{.}{.}{-1}}

% Text layout (margins)
\topmargin 0.0cm
\oddsidemargin 0.5cm
\evensidemargin 0.5cm
\textwidth 16cm 
\textheight 21cm

% double line spacing
\linespread{1.6}

% paragraph indent
\setlength{\parindent}{0.5in}

% font for math
\usepackage{mathpazo}
\DeclareSymbolFont{letters}{OML}{ztmcm}{m}{it}
\DeclareSymbolFontAlphabet{\mathnormal}{letters}

\definecolor{fgcolor}{rgb}{0.2, 0.2, 0.2}


%%%%%%%%%%%%%%%%
%---DOCUMENT---%
%%%%%%%%%%%%%%%%
\begin{document}

% consistent edits between manuscript and rebuttal letter
\input{thesis.xtr}

\setcounter{page}{1}

\bigskip
\medskip
\noindent{\Large 
Response 
}
\bigskip


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%---SUGGESTED SIMULATION SCENARIOS---%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\section*{Simulations}

\begin{itemize}

\subsection*{David Posada}

\item {
{\it
\textbf{
Simulate a number of epochs that do not correspond with the assumed ones.
}% END: textbf
}% END: it
}% END: item
\\
FB: Perhaps two sets of simulations here: \\
* simulate 2 epochs - inference under 1 \\
* simulate 3 - inference for 2 \\
Check the variance and bias

\item {
{\it
\textbf{
Simulating under a non-epoch model, for example where different process can apply to contemporaneous branches, would be fundamental to understand the robustness of the model.
}% END: textbf
}% END: it
}% END: item
\\
FB: I think he means we did simulations with an ultrametric tree, but we simulated and recovered the values under the same (epoch) model.
For that I could set up a scenario where data is generated under e.g. HKY(1.0) with an underlying ultrametric topology and then we recover two epochs, placing the transition time somewhere arbitrary.




\item {
{\it
\textbf{
Here it would have been interesting to see the correspondence between simulated and estimated tMRCAs when the true model has one epoch with omega=0.5 (or even better, omega equal to the weighted mean of the n-epoch model).
}% END: textbf
}% END: it
}% END: item
\\
FB: I think here we could repeat the study from the $\pi$BUSS paper but simulate under a homogenous model. We indeed could set the $\omega$ parameter to a weighted mean of the values we used previously (1.0, 0.2, 0.1, 0.02, 0.01, 0.002, 0.001), as weights we could use epoch length, where for last epoch it's tMRCa - T[n]. At least that's how I understand this comment.

\subsection*{Stijn Vansteelandt}

\item {
{\it
\textbf{
It would also have been nice to understand better whether the use of the
epoch model (versus not) has an essential impact on tree reconstruction.
}% END: textbf
}% END: it
}% END: item
\\
FB: I don't think it will have an impact, but to proove it simulations would be adequate.


\item {
{\it
\textbf{
How mislead one could be if the data were generated under an epoch model, but analyzed under a simpler model.
}% END: textbf
}% END: it
}% END: item
\\
I think this comment is asking basically the same thing as David Posada's first suggested simulation.


\end{itemize}


%TODO: I'm including only reviewers which had some major/minor comments

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%--- COMMENTS BY STIJN VANSTEELANDT---%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\section*{Comments by Stijn Vansteelandt}

\begin{itemize}

% 1
\item {
{\it
\textbf{
The thesis starts with a very nice introduction, which I think is very insightful for the reader who is not familiar with phylogenetic and/or Bayesian inference. 
In spite of this, I believe that the readability of the later chapters would be further improved if the introduction:
}% END: textbf
}% END: it
}% END: item

\begin{itemize}

% 1a
\item {
{\it
\textbf{
Briefly described the evolutionary models, such as GY94, that are later used.
}% END: textbf
}% END: it
}% END: item
\\
Done, there is now a \emph{Markov models of codon evolution} section in the introductory chapter.


% 1b
\item {
{\it
\textbf{
Briefly described how to accommodate time-stamped sequences, which is now considered in the analyses in Chapter 2, but not really discussed.
}% END: textbf
}% END: it
}% END: item
\\
AAA




% 1c
\item {
{\it
\textbf{
Give a bit more detail on phylogeography, e.g. how do the models look like and what is the connection between phylogeography and epoch models (some connection is described in Chapter 2, but remains a bit vague).
}% END: textbf
}% END: it
}% END: item
\\
AAA

\end{itemize}

% 2
\item {
{\it
\textbf{
It is sometimes unclear in Chapter 2 whether the analysis is performed under a fixed evolutionary tree or not. 
This is especially so in the evaluation of Bayes Factors in Table 2.3. 
It would also have been nice to understand better whether the use of the epoch model (versus not) has an essential impact on tree reconstruction. 
Somewhat related, it would have been nice to learn from the simulation study how mislead one could be if the data were generated under an epoch model, but analyzed under a simpler model. 
I am not asking for additional analyses here, but merely for additional clarification or insight.
}% END: textbf
}% END: it
}% END: item
\\
AAA

% 3
\item {
{\it
\textbf{
P6, lines 5-7: this is not correctly phrased as the parameter values do not have an associated probability in the frequentist likelihood framework.
}% END: textbf
}% END: it
}% END: item
\\
AAA

% 4
\item {
{\it
\textbf{
Expression (1.12) is confusing in that indices i and j appear in the lefthand side, but not in the righthand side. 
Likewise in expression (1.13).
}% END: textbf
}% END: it
}% END: item
\\
AAA

% 5
\item {
{\it
\textbf{
P20, 2nd display: an equality sign is missing on the 2nd line.
}% END: textbf
}% END: it
}% END: item
\\
AAA

% 6
\item {
{\it
\textbf{
I would find it useful to see a brief discussion on the plausibility of (1.22).
}% END: textbf
}% END: it
}% END: item
\\
AAA

% 7
\item {
{\it
\textbf{
If I am right that matrix P(t) on the bottom of page 22 is four-dimensional, then why not write it in full, which will be less confusing?
}% END: textbf
}% END: it
}% END: item
\\
AAA

% 8
\item {
{\it
\textbf{
P25, the statement at the end of the penultimate paragraph (`most plausible hypothesis for the data') is vague.
}% END: textbf
}% END: it
}% END: item
\\
AAA

% 9
\item {
{\it
\textbf{
P28, I understand what T(n) is, but would nevertheless find it helpful if it were stated more explicitly.
}% END: textbf
}% END: it
}% END: item
\\
AAA

% 10
\item {
{\it
\textbf{
P31, If the phrase `proper prior' is used, then perhaps it should be defined.
}% END: textbf
}% END: it
}% END: item
\\
AAA



% 11
\item {
{\it
\textbf{
P32, Since the author assumes that the reader may have no prior knowledge of Bayesian inference, I believe the meaning of Figure 1.11 should be explained in more detail.
}% END: textbf
}% END: it
}% END: item
\\
AAA


% 12
\item {
{\it
\textbf{
P33,. I think the notation $\theta[t]$ has not been defined.
}% END: textbf
}% END: it
}% END: item
\\
AAA


% 13
\item {
{\it
\textbf{
There is a typo in the lefthand side of (1.37).
}% END: textbf
}% END: it
}% END: item
\\
AAA


% 14
\item {
{\it
\textbf{
I would find it useful to get a bit more detail how a model for the likelihood of Y and for the prior of $\Phi$ might look like.
}% END: textbf
}% END: it
}% END: item
\\
AAA


% 15
\item {
{\it
\textbf{
P43. What is compositional heterogeneity?
}% END: textbf
}% END: it
}% END: item
\\
AAA


% 16
\item {
{\it
\textbf{
P44. I would find it useful to receive a bit more guidance concerning the connection to phylogeographic inference. 
Also in Section 2.4.3, the description of the proposal by Bahl et al. (2011) is too brief for me to appreciate why the epoch model is more appropriate.
}% END: textbf
}% END: it
}% END: item
\\
AAA


% 17
\item {
{\it
\textbf{
Display (2.5): I believe this display implicitly assumes that the time T1 is a priori known. 
It would be good to state this explicitly.
}% END: textbf
}% END: it
}% END: item
\\
AAA


% 18
\item {
{\it
\textbf{
P57. The sentence `To determine the posterior odds, ... of that model' is vague.
}% END: textbf
}% END: it
}% END: item
\\
AAA


% 19
\item {
{\it
\textbf{
P85. Riemannian -> Riemann
}% END: textbf
}% END: it
}% END: item
\\
Corrected.


% 20
\item {
{\it
\textbf{
P86. The last sentence of Section 5.1.3 is not sufficiently clear. 
Please make clear why the use of mixed effects models could allow this.
}% END: textbf
}% END: it
}% END: item
\\
AAA


% 21
\item {
{\it
\textbf{
P88. It would be useful to receive a bit more detail what is meant by `infinite hidden Markov Model'.
}% END: textbf
}% END: it
}% END: item
\\
AAA



\end{itemize}


%%%%%%%%%%%%%%%%%%%%
%--- COMMENTS---%
%%%%%%%%%%%%%%%%%%%%
\section*{Comments by David Posada}

\begin{itemize}

% 1
\item {
{\it
\textbf{
I missed a brief primer of Bayesian model selection in the introductory chapter.
Techniques such Bayes factors, path sampling or stepping-stone sampling are used throughout but never explained. 
This should be very easy to fix.
}% END: textbf
}% END: it
}% END: item
\\
There is now a subsection called \emph{Bayesian model selection} at the end of Chapter 1.



% 2
\item {
{\it
\textbf{
In the simulations in Chapter 2 you assume you know the true epoch times, but this information is rarely available in real life. 
Simulation 1 addresses the impact of model misspecification partially, as the true process does not have epochs. 
It could be a good idea to further check the impact of model misspecification, in particular simulating a number of epochs that do not correspond with the assumed ones (you later mention this idea in the final chapter but do not end up testing its potential biasing effect). 
Also, simulating under a non-epoch model - for example where different process can apply to contemporaneous branches, would be fundamental to understand the robustness of the model. 
This does not need to be done before the defense, but it could be a nice extension for the future.
}% END: textbf
}% END: it
}% END: item
\\
AAA

% 3
\item {
{\it
\textbf{
In general, there is some repetitive text, sometimes with identical phrases/paragraphs, between the introduction and some of the other chapters. 
On the other hand, some relevant aspects of the research are discussed in the general discussion in Chapter 5, instead of in the corresponding `research' chapter. 
This is true in particular for Chapter 3, were a very concise description of SPREAD is provided. 
I understand these redundancies and fragmentations are a byproduct of reproducing papers as chapters, but however these issues have been corrected in some occasions. 
For example, chapter 2 is not identical to the SysBio paper. 
I suggest you try a bit to further consolidate the different chapters.
}% END: textbf
}% END: it
}% END: item
\\
AAA

% 4
\item {
{\it
\textbf{
The simulation software piBUSS described in Chapter 4 can be very useful, in particular in regards to its connectivity with BEAST. 
The possibility of simulating discrete phylogeographic traits is very convenient. 
Here it would have been interesting to see, for the sake of comparison and interpretation, the correspondence between simulated and estimated tMRCAs when the true model has one epoch with omega=0.5 (or even better, omega equal to the weighted mean of the n-epoch model). 
This does not need to be done before the defense, but it could be a nice extension for the future.
}% END: textbf
}% END: it
}% END: item
\\
AAA


% 5
\item {
{\it
\textbf{
Page 14: I would say that the term `Big Data' does not apply to sequence alignments. 
It might do for NGS files etc., but not for the type of data that BEAST uses as input. 
Not yet. 
Moreover, you focus on HPC architectures like GPU and multicores, which are in some senses opposite to Big Data architectures (i.e., Hadoop Map Reduce).
}% END: textbf
}% END: it
}% END: item
\\
Rephrased to avoid term:

\begin{quote}
\myeditDPfive
\end{quote}


% 6
\item {
{\it
\textbf{
Page 25, 31: likelihood should be $P(X| \theta)$, not $P(\theta | X)$.
}% END: textbf
}% END: it
}% END: item
\\
Corrected.


% 7
\item {
{\it
\textbf{
Page 34: on the left-term of equation 1.37, the denominator should be
$P(\theta[t]|X) q(\theta^{*} | \theta[t])$
not 
$P(\theta[t]|X) q(\theta[t] | \theta^{*})$
}% END: textbf
}% END: it
}% END: item
\\
Corrected.


% 8
\item {
{\it
\textbf{
Page 52: according to the legend in Figure 2.3, a two-epoch plot is missed.
}% END: textbf
}% END: it
}% END: item
\\
Hmm I think it wasn't clear that it's all on one plot. Needs re-phrasing:

\begin{quote}
{\bf  Epoch simulation scenarios on an influenza A maximum clade credibility tree topology.} 
In the two-epoch example illustrated by the top alternating background
% at the top, 
the transition time is set at $t_{1}=7$, creating two epochs with substitution processes governed by infinitesimal rate matrices $\mathbf{Q_{1}}$ and $\mathbf{Q_{2}}$ respectively, separated by the light and dark grey areas and the dotted line.
In the three-epoch example illustrated 
by the bottom alternating background
% at the bottom, 
transition times are put at $t_{1}=7$ and $t_{2}=15$, creating three epochs with substitution processes governed by infinitesimal rate matrices $\mathbf{Q_{1}}$, $\mathbf{Q_{2}}$ and then again $\mathbf{Q_{1}}$, as indicated by the alternating dark and light areas and the dotted lines.
\end{quote}











\end{itemize}











%\bibliography{epochrefs}

\end{document}
